ee/agent/storage/bbolt/backup.go

```diff
@@ -0,0 +1,173 @@
+	}
+
+	// Confirm file exists and is nonempty
+	if exists, err := launcher.NonEmptyFileExists(backupLocation); !exists {
+		return fmt.Errorf("backup succeeded, but nonempty file does not exist at %s", backupLocation)
+	} else if err != nil {
+		return fmt.Errorf("backup succeeded, but error checking if file was created at %s: %w", backupLocation, err)
+	}
+
+	// Log success
+	d.slogger.Log(context.TODO(), slog.LevelDebug,
+		"took backup",
+		"backup_location", backupLocation,
+	)
+
+	return nil
+}
+
+// UseBackupDbIfNeeded falls back to the backup database IFF the original database does not exist
+// and the backup does. In this case, it renames the backup database to the expected filename
+// launcher.db.
+func UseBackupDbIfNeeded(rootDir string, slogger *slog.Logger) {
+	// Check first to see if the regular database exists
+	originalDbLocation := LauncherDbLocation(rootDir)
+	if originalDbExists, err := launcher.NonEmptyFileExists(originalDbLocation); originalDbExists {
+		// DB exists -- we should use that
+		slogger.Log(context.TODO(), slog.LevelDebug,
+			"launcher.db exists, no need to use backup",
+			"db_location", originalDbLocation,
+		)
+		return
+	} else if err != nil {
+		// Can't determine whether the db exists -- err on the side of not replacing it
+		slogger.Log(context.TODO(), slog.LevelWarn,
+			"could not determine whether original launcher db exists, not going to use backup",
+			"err", err,
+		)
+		return
+	}
+
+	// Launcher DB doesn't exist -- check to see if the backup does
+	backupLocation := BackupLauncherDbLocation(rootDir)
+	backupDbExists, err := launcher.NonEmptyFileExists(backupLocation)
+	if !backupDbExists {
+		// Backup DB doesn't exist either -- this is likely a fresh install.
+		// Nothing to do here; launcher should create a new DB.
+		slogger.Log(context.TODO(), slog.LevelInfo,
+			"both launcher db and backup db do not exist -- likely a fresh install",
+		)
+		return
+	}
+	if err != nil {
+		// Couldn't determine if the backup DB exists -- let launcher create a new DB instead.
+		slogger.Log(context.TODO(), slog.LevelWarn,
+			"could not determine whether backup launcher db exists, not going to use backup",
+			"err", err,
+		)
+		return
+	}
+
+	// The backup database exists, and the original one does not. Rename the backup
+	// to the original so we can use it.
+	if err := os.Rename(backupLocation, originalDbLocation); err != nil {
+		slogger.Log(context.TODO(), slog.LevelWarn,
+			"could not rename backup db",
+			"backup_location", backupLocation,
+			"original_location", originalDbLocation,
+			"err", err,
+		)
+		return
+	}
+	slogger.Log(context.TODO(), slog.LevelInfo,
+		"original db does not exist and backup does -- using backup db",
+		"backup_location", backupLocation,
+		"original_location", originalDbLocation,
+	)
+}
+
+func LauncherDbLocation(rootDir string) string {
+	return filepath.Join(rootDir, "launcher.db")
+}
+
+func BackupLauncherDbLocation(rootDir string) string {
```

Comment:
I wasn't super worried about that because of how the access remains relatively restricted in this implementation --

* On launcher startup, before any rungroups execute, launcher may rename `launcher.db.bak` to `launcher.db`; nothing else is accessing either db at this time
* After rungroups begin to execute, this actor is the only thing touching `launcher.db.bak` -- nothing else will read from it or write to it
* (Except for the flare checkup -- and data inconsistency probably isn't a huge deal there, and I'm not sure how much we even care about having that checkup anyway)
* (There's also remote uninstall, but that just wants to delete the file)

Is there something I'm missing?

I'm happy to keep N and rotate if we feel like that's a safer option, regardless.

Reply:
I added rotation in https://github.com/kolide/launcher/pull/1755/commits/15183fa45690392898d559137063fc8e6565b9c3 -- lmk what you're thinking!